demos/tutorial/new_bc_2.cpp

```diff
@@ -0,0 +1,308 @@
+auto make_region(samurai::subset_operator<F, CT...> region)
+{
+    return std::make_unique<SetRegion<dim, samurai::subset_operator<F, CT...>>>(region);
+}
+
+template<std::size_t dim, class Func>
+auto make_region(Func&& func)
+{
+    return std::make_unique<CoordsRegion<dim>>(std::forward<Func>(func));
+}
+
+// Bc
+/////////////////////////////////////////////////////////////////////////////////////////
+template<std::size_t dim, class T, std::size_t size>
+struct Bc
+{
+    virtual ~Bc() = default;
+
+    template<class Bcvalue>
+    Bc(const Bcvalue& bcv)
+
+    {
+        p_bcvalue = std::make_unique<Bcvalue>(bcv);
+        p_bcregion = std::make_unique<Everywhere<dim>>();
+    }
+
+    Bc(const Bc& bc)
+    : p_bcvalue(bc.p_bcvalue->clone())
+    , p_bcregion(bc.p_bcregion->clone())
+    {
+    }
+
+    template<class Region>
+    auto on(const Region& region)
+    {
+        p_bcregion = make_region<dim>(region);
+        return *this;
+    }
+
+    auto get_lca(const samurai::LevelCellArray<dim>& mesh)
+    {
+        return p_bcregion->apply(mesh);
+    }
+
+    std::unique_ptr<BcValue<dim, T, size>> p_bcvalue;
+    std::unique_ptr<BcRegion<dim>> p_bcregion;
+};
+
+
+template<std::size_t dim, class T, std::size_t size>
+struct Dirichlet: public Bc<dim, T, size>
+{
+    using Bc<dim, T, size>::Bc;
+};
+
+template<std::size_t dim, class T, std::size_t size>
+struct Neumann: public Bc<dim, T, size>
+{
+    using Bc<dim, T, size>::Bc;
+};
+
+template<std::size_t dim, class T, std::size_t size>
+struct Robin: public Bc<dim, T, size>
+{
+    using Bc<dim, T, size>::Bc;
+};
+
+template<std::size_t dim, class T, std::size_t size>
+struct Field
+{
+    Field(){}
+
+    template<class Bc_derived>
+    void attach(const Bc_derived& bc)
+    {
+        //p_bc.push_back(std::unique_ptr<Bc_derived>(bc.clone()));
+        p_bc.push_back(std::make_unique<Bc_derived>(bc));
+    }
+    std::vector<std::unique_ptr<Bc<dim, T, size>>> p_bc;
+};
+
+template<class Field, std::size_t dim=Field::dim, class T = typename Field::type, std::size_t size=Field::size>
+void apply(const Dirichlet<dim, T, size>& dirichlet, Field& field)
+{
+    std::cout << "Dirichlet" << std::endl;
+    std::cout << dirichlet.p_bcvalue->apply({1, 2})[0] << std::endl;
+}
+
+template<class Field, std::size_t dim=Field::dim, class T = typename Field::type, std::size_t size=Field::size>
+void apply(const Neumann<dim, T, size>& neumann, Field& field)
+{
+    std::cout << "Neumann" << std::endl;
+    std::cout << neumann.p_bcvalue->apply({1, 2})[0] << std::endl;
+}
+
+int main()
+{
+    constexpr std::size_t dim = 2;
+    samurai::Box<double, dim> box = {{0, 0}, {1, 1}};
+    samurai::LevelCellArray<dim> lca = {2, box};
+
+    Field<2, double, 3> f;
+    Dirichlet<2, double, 3> dirichlet(ConstantBc<2, double, 3>{4});
+    dirichlet.on([](auto& coords)
+    {
+        return (coords[0] >= .25 && coords[0] <= .75);
+    });
+
+    f.attach(dirichlet);
```

Comment:
The user code is still too complex to me.
One should be able to write something like 
`auto bc = dirichlet(4).on(...);`
without specifying ConstantBc or FunctionBc.

It would also be nice that `attach()` return *bc so that we can also write
`f.attach(dirichlet(4)).on(...)`